File: 20. Valid Parentheses.java
class Solution {
    private HashMap<Character, Character>mappings;
    
    public Solution() {
        this.mappings = new HashMap<>();
        this.mappings.put(')','(');
        this.mappings.put(']','[');
        this.mappings.put('}','{');
    }
    
    
    public boolean isValid(String s) {
        Stack<Character> stack = new Stack<Character>();
        for (int i = 0; i < s.length(); i++) {
            char cur = s.charAt(i);
            if (cur == '(' || cur == '[' || cur == '{') {
                stack.push(cur);
            }
            else if (stack.isEmpty()) 
                return false;
            
            else if (mappings.get(cur) == stack.peek()) {
                stack.pop();
            }
            else {
                return false;
            }
        }
        return stack.isEmpty();
    }
    
    
}

// Time complexity: O(n)
// Space complexity: O(n)

File: amazon/python/20. Valid Parentheses.py
class Solution(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        stack = []
        if len(s) == 0:
            return True
        
        
        for char in s:
            if char in ["(","[","{"]:
                stack.append(char)
            elif len(stack) == 0:
                return False
            else:
                last = stack.pop()
                if not ((char == ")" and last == "(") or (char == "]" and last == "[") or (char == "}" and last == "{")):
                    return False
                             
        return len(stack) == 0
    
    ## Time: O(n)
    ## Space: O(n)
                
                
